_targets.R

```diff
@@ -0,0 +1,27 @@
+library(targets)
+
+options(tidyverse.quiet = TRUE)
+tar_option_set(packages = c('tidyverse', 'lubridate', 'dataRetrieval'))
+
+source("1_fetch.R")
+
+# Define the temporal extent of our data pull
+start_date <- '2015-01-01'
+end_date <- '2015-12-31'
+
+# The spatial extent of our data pull is the Delaware River Basin (DRB), so 
+# here we will define the minor HUCs (hydrologic unit codes) that make up the 
+# DRB to pass to data download functions in dataRetrieval. 
+# The DRB is represented by subregion '0204' (https://water.usgs.gov/GIS/huc_name.html)
+drb_huc8s <- c('02040101','02040102','02040103','02040104','02040105','02040106',
+               '02040201','02040202','02040203','02040204','02040205','02040206','02040207')
```

Comment:
> How did you make the decision to use HUCs rather than a bounding box? Did this come from the existing DRB scripts?

Good question! My thought was that the bounding box would include some areas outside of the Delaware River Basin, and so would potentially return more data than we'd want if we were only interested in the DRB. I know that `huc` is an attribute included in WQP, so I figured we could use that. _However_, I can certainly imagine cases where the HUC8 field isn't populated in WQP and so wouldn't get returned. Based on this advice from Jordan and Sam, I propose that we use a bounding box instead of the HUC8's here, and perhaps include some documentation that formalizes that reasoning.

Reply:
Agreed :)
Another reason to avoid HUC in this case since the example will be pretty generic and re-usable outside of the immediate project needs is that HUC is kind of jargony and hard for lots of users to connect with. My pattern for this type of thing in a past project would be to
1) make a holistic grid of bboxes
2) take my shape that I'm interested in (in your case, the DRB outline, perhaps buffered?)
3) query all bboxes that have _any_ overlap with the shape for sites with data
4) combine all sites from 3 and filter out any sites outside of the shape (point-in-poly analysis)
5) set up a set of calls to make for the data, using some kind of grouping. 

☝️ one reason to create a big set of grids first is that it makes it easier to expand your region of interest without having to re-pull tiles that have data. Not saying this is the best way to do it, but another data point to consider :)

Reply:
Thanks for this tip, @jread-usgs! I agree that HUC's are jargony and I like that the grid-based approach is potentially more amenable to scaling depending on the use case. I've created an issue to implement the grid-based approach in our pipeline. One thing I plan to keep an eye on, though, is the evenness of the data that gets returned - are the resulting data pulls roughly equal across boxes, or will we want to further partition the site numbers that get returned from our initial grid-based inventory?

Reply:
Yes, you are right that the amount of sites (and data) within each grid cell varies a lot. Years ago, the WQP was capable of returning _all sites_ meeting criteria without specifying any spatial filter, but that kind of query now times out so it is necessary to use any one of bbox, state, county, HUC etc. 

My past workflows of this type use a much smaller number of grid cells to make the site query than the number of data calls. For example, I pull temperature sites using a 2° or 5° lat/lon cell, depending on whether I am limiting to all sites (2°) or just lake sites (5°). Since those are pretty large, you'd probably only have a handful of cells in your site query for the DRB. The data query ends up with way more calls because you are then getting the time series data, and I separate those out like others have done and shared with this team by using two thresdholds: number of sites within a single data call can't exceed X and approximate number of total records for all of the sites in the call can't exceed Y, whichever threshold is hit first. I've made this split before by ranking all sites according to `resultCount` and partitioning from there, but a slight variation on that would make things scalable without as much repetition in the calls. If instead you grouped the sites and partitions according to grid cell, then your partitions in the DRB don't get re-scrambled if you add a cell outside of that initial pull (and thus, only the new sites would get pulled).